design-proposals/2023-02-08-code-organisation.md

```diff
@@ -0,0 +1,137 @@
+  - `platform/windows`
+  - `platform/qt`
+- `test`: common (headless) tests for all platforms; disabled only in exceptional cases if some features are not applicable due to the nature of a specific platform
+
+The core itself ideally should have no binaries unless they are really platform-independent (e.g. headless rendering).
+
+### Supported platforms
+
+We can split the platforms into two categories:
+
+- **Native platforms / operating systems**
+  - Windows
+  - Linux
+  - Darwin and its derivatives (macOS, iOS, tvOS, ...)
+  - Android
+- **Cross-platform engines**
+  - Node.js
+  - Qt
+  - GLFW
+
+### Build infrastructure
+
+The core should be exclusively built by `CMake`. All currently supported platforms are supported by it. No wrapper script or `Makefile` should be used. Can export `Xcode` projects if needed.
+
+The functionality part of the library (under `src`) should yield a separate object library. Similarly the platform-specific part (under `platform`) should be separate. Each 3rd-party dependency should also build as its own object library if distributed together with MapLibre. Static libraries may only be used as backwards compatibility for older platforms that the community may be interested in.
+
+The code should be C++17 or later (unless required otherwise by platform-specific integration).
+
+Compilation flags should ideally be the same on one native platform / operating system. This would allow building multiple platform-specific parts in parallel and would speed-up the development.
+
+Final products of the core would be two or more object libraries to be directly used by the SDKs or one static or dynamic library for binary distribution. *Note: At this point it is not clear if all platforms need binary distribution independent of the SDK that will be provided for that platform.*
+
+Examples of binary distributions could be:
+
+- Dynamic libraries for Windows, Linux and Android.
+- Dynamic and static frameworks for macOS and iOS.
+- Cross-platform engines will probably build the core as part of the SDK and no direct core binaries are needed (as written above this may also apply for native platforms).
+
+### Testing
+
+All testing should be done for all platforms at the same time and as complete as possible. Basic rendering tests should be done headlessly for all platforms.
+
+If needed for specific platforms (iOS comes to mind), a dedicated test app could be made to run the necessary rendering tests. This app should implement the minimal functionality that does not include the actual full-fledged integration of MapLibre (e.g. gestures, widgets, ...).
+
+Testing should be done by `CTest` if the platform allows (may not be possible for all integration tests on mobile devices).
+
+### Release cycle
+
+The core should have an independent release cycle. At this point we imagine SDKs to be much more actively developed. This also allows bigger refactoring to take place without affecting SDKs (e.g. Metal migration).
```

Comment:
Correct me if I am wrong, but if I understand you correctly, what you are pointing out here is that we can make a big changes to the core and then each individual platform can adopt those changes at their own pace at a later time. I agree there is a benefit to this, but at the same time I think that core changes have no benefit on their own if the supported platforms are not modified alongside those changes. The big benefit right now is that if you make a change that breaks a platform in the core, we know it immediately because CI will fail (in theory). Updating a platform that is pinned to a specific core version will potentially be much harder with a indepdenent release cycle, because all changes would need to be traced down.

An independent release cycle without separate repos would not have my preference. That would just be very confusing in my eyes.

I think it is possible to make bigger refactorings without affecting the SDKs with a monorepo by using feature flags and the likes.

Reply:
If the public-facing API of the core stays the same, there is no need to explicitly release SDKs. I guess from practical point of view this will still happen, but may not happen as often.

A clear separation would also help clients reverting to older versions of core temporarily if bugs are found.

Reply:
Interesting conversation. If the public facing core API does not change, the platforms need not to be changed. However, can we guarantee for platforms like Android static linking won't be required with core?

As in should the build artifact from core also contain source?

Reply:
In principle none of the platforms would require static linking, I currently do use dynamic libraries on Android, and static on iOS. Usually static linking reduces the size a bit though.